controller/src/main/java/io/pravega/controller/store/stream/ZookeeperBucketStore.java

```diff
@@ -0,0 +1,261 @@
+    private static final String ROOT_PATH = "/";
+    private static final String OWNERSHIP_CHILD_PATH = "ownership";
+    @Getter
+    private final int bucketCount;
+    private final ConcurrentMap<ServiceType, PathChildrenCache> bucketOwnershipCacheMap;
+    private final ConcurrentMap<String, PathChildrenCache> bucketCacheMap;
+    private final ZKStoreHelper storeHelper;
+
+    protected ZookeeperBucketStore(int bucketCount, CuratorFramework client, Executor executor) {
+        this.bucketCount = bucketCount;
+        this.bucketCacheMap = new ConcurrentHashMap<>();
+        this.bucketOwnershipCacheMap = new ConcurrentHashMap<>();
+        storeHelper = new ZKStoreHelper(client, executor);
+    }
+
+    @Override
+    public void registerBucketOwnershipListener(ServiceType serviceType, BucketOwnershipListener listener) {
+        Preconditions.checkNotNull(listener);
+
+        String bucketRootPath = ZKPaths.makePath(ROOT_PATH, serviceType.getName());
+        PathChildrenCacheListener bucketListener = (client, event) -> {
+            switch (event.getType()) {
+                case CHILD_ADDED:
+                    // no action required
+                    break;
+                case CHILD_REMOVED:
+                    int bucketId = Integer.parseInt(ZKPaths.getNodeFromPath(event.getData().getPath()));
+                    listener.notify(new BucketNotification(bucketId, BucketNotification.NotificationType.BucketAvailable));
+                    break;
+                case CONNECTION_LOST:
+                    listener.notify(new BucketNotification(Integer.MIN_VALUE, BucketNotification.NotificationType.ConnectivityError));
+                    break;
+                default:
+                    log.warn("Received unknown event {} on bucket root {} ", event.getType(), serviceType);
+            }
+        };
+
+        String bucketOwnershipPath = ZKPaths.makePath(bucketRootPath, OWNERSHIP_CHILD_PATH);
+        bucketOwnershipCacheMap.computeIfAbsent(serviceType, 
+                x -> {
+                    PathChildrenCache pathChildrenCache = new PathChildrenCache(storeHelper.getClient(), 
+                            bucketOwnershipPath, true);
+
+                    pathChildrenCache.getListenable().addListener(bucketListener);
+                    try {
+                        pathChildrenCache.start(PathChildrenCache.StartMode.BUILD_INITIAL_CACHE);
+                    } catch (Exception e) {
+                        log.error("Starting ownership listener for service {} threw exception {}", serviceType, e);
+                        throw Lombok.sneakyThrow(e);
+                    }
+                    log.info("bucket ownership listener registered on bucket root {}", serviceType);
+
+                    return pathChildrenCache;
+                });
+        }
+
+    @Override
+    public synchronized void unregisterBucketOwnershipListener(ServiceType serviceType) {
```

Comment:
Why are these methods synchronized? It appears that IO could occur inside of this lock.